### PR TITLE
Fixed some bugs in codegenerator plugin

### PR DESCRIPTION
**Pull request by rakshitkr**
Signed-off-by: Rakshit Krishnappa Ravi <rakshitkrishnapparavi@gmail.com>

- correct the package names of SecretKeySpec, SecretKey, PBEKeySpec
- fixed NPE in setUpTemplateClass method
- removed unwanted logger in isSubType method
- fixed a bug in collectParameterTypes method
- removed JDialog to increase the runtime further
- refactored setUpTemplateClass method to improve runtime

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:
* Eclipse Version:
* Java Version:
* OS:

# Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by rakshitkr** [comment]
@shahrzadav I've fixed all the bugs in the latest codegenerator plugin. Feel free to test it once and let me know if you face any issues.

File: plugins/de.cognicrypt.codegenerator/src/main/java/de/cognicrypt/codegenerator/generator/CrySLBasedCodeGenerator.java [hunk 1]
```diff
@@ -1232,6 +1232,7 @@ public GeneratorClass setUpTemplateClass(String pathToTemplateFile) {
 			Map<SimpleName, CrySLObject> postCGVars = new HashMap<SimpleName, CrySLObject>();
 
 			List<CodeGenCrySLRule> rules = new ArrayList<CodeGenCrySLRule>();
+			List<CrySLRule> rulesFromSootRunner = SootRunner.getRules(getDeveloperProject().project);
```
Comment (schlichtig) [inline plugins/de.cognicrypt.codegenerator/src/main/java/de/cognicrypt/codegenerator/generator/CrySLBasedCodeGenerator.java:1236]: Why did you move this up?

**Comment by schlichtig** [comment]
@rakshitkr Jenkins build is failing. Please check this.

File: plugins/de.cognicrypt.codegenerator/src/main/java/de/cognicrypt/codegenerator/generator/CrySLBasedCodeGenerator.java [hunk 2]
```diff
@@ -683,11 +683,11 @@ private CrySLMethod fetchCorrespondingMethod(Entry<CrySLPredicate, Entry<CrySLRu
 			} else {
 				try {
 					methodParameter[i] = Class.forName(parameter.getValue());
-					i++;
 				} catch (ClassNotFoundException e) {
 					Activator.getDefault().logError(e, "No class found for type: " + parameter.getValue().toString());
 				}
 			}
+			i++;
```
Comment (schlichtig) [inline plugins/de.cognicrypt.codegenerator/src/main/java/de/cognicrypt/codegenerator/generator/CrySLBasedCodeGenerator.java:691]: Why did you move this out of the try-block?

File: plugins/de.cognicrypt.codegenerator/src/main/java/de/cognicrypt/codegenerator/wizard/AltConfigWizard.java [hunk 3]
```diff
@@ -238,12 +238,6 @@ public boolean performFinish() {
 		IResource targetFile = (IResource) currentPage.getSelectedResource().getFirstElement();
 
 		String taskName = selectedTask.getName();
-		JOptionPane optionPane = new JOptionPane("CogniCrypt is now generating code that implements " + selectedTask.getDescription() + "\ninto file " + ((targetFile != null)
```
Comment (schlichtig) [inline plugins/de.cognicrypt.codegenerator/src/main/java/de/cognicrypt/codegenerator/wizard/AltConfigWizard.java:241]: How much does this affect runtime?

File: plugins/de.cognicrypt.core/src/de/cognicrypt/utils/Utils.java [hunk 4]
```diff
@@ -295,7 +295,7 @@ public static boolean isSubType(String typeOne, String typeTwo) {
 				subTypes = Class.forName(typeOne).isAssignableFrom(Class.forName(typeTwo));
 			}
 			catch (ClassNotFoundException e) {
-				Activator.getDefault().logError(e);
+				// It's fine if above throws a ClassNotFoundException
```
Comment (schlichtig) [inline plugins/de.cognicrypt.core/src/de/cognicrypt/utils/Utils.java:298]: Why? Where is the Error handled above?

**Review by schlichtig — CHANGES_REQUESTED** [review]
Please che the comments and why Jenkins fails.

File: plugins/de.cognicrypt.codegenerator/src/main/java/de/cognicrypt/codegenerator/generator/CrySLBasedCodeGenerator.java (hunk 1 above)
Comment (rakshitkr) [reply plugins/de.cognicrypt.codegenerator/src/main/java/de/cognicrypt/codegenerator/generator/CrySLBasedCodeGenerator.java:1236]: Because the `visit` method is called multiple times, the rules are retrieved repeatedly if `SootRunner.getRules` is inside this method.

File: plugins/de.cognicrypt.codegenerator/src/main/java/de/cognicrypt/codegenerator/generator/CrySLBasedCodeGenerator.java (hunk 2 above)
Comment (rakshitkr) [reply plugins/de.cognicrypt.codegenerator/src/main/java/de/cognicrypt/codegenerator/generator/CrySLBasedCodeGenerator.java:691]: This is a bug. `i` never gets incremented for primitive type parameters.

File: plugins/de.cognicrypt.codegenerator/src/main/java/de/cognicrypt/codegenerator/wizard/AltConfigWizard.java (hunk 3 above)
Comment (rakshitkr) [reply plugins/de.cognicrypt.codegenerator/src/main/java/de/cognicrypt/codegenerator/wizard/AltConfigWizard.java:241]: The actually issue was that the dialog appears in the background (not consistently) and doesn't close even after the code generation process is completed. Have added it back in 523d910 and have set `visibility` to false after generation is finished.

File: plugins/de.cognicrypt.core/src/de/cognicrypt/utils/Utils.java (hunk 4 above)
Comment (rakshitkr) [reply plugins/de.cognicrypt.core/src/de/cognicrypt/utils/Utils.java:298]: The logic works like "If both the type strings `typeOne` and `typeTwo` are convertible to classes, then check for subtype relation. Otherwise they are assumed to be primitive or primitive arrays and existing `subTypes` value calculated previously is returned".

**Comment by rakshitkr** [comment]
@shahrzadav Is the code generator working for all cryptographic use cases after the fix that I've provided for your issue?
@schlichtig Is this PR ready to be merged?

**Comment by shahrzadav** [comment]
@rakshitkr thanks for the fixes. yes it is working fine for me and I can generate the desired code in Encryption use case. Also fixes for SecureRandom class resolved my problem.
